File: api/research_config.py
"""Research configuration management with user-specific settings."""

import os
from typing import Dict, Any, Optional
from langchain_core.runnables import RunnableConfig

from open_deep_research.configuration import Configuration


class ResearchConfigManager:
    """Manages research configuration for users and sessions."""
    
    def __init__(self):
        self.default_config = self._load_default_config()
    
    def _load_default_config(self) -> Dict[str, Any]:
        """Load default configuration from environment variables."""
        return {
            "max_structured_output_retries": int(os.getenv("MAX_STRUCTURED_OUTPUT_RETRIES", "3")),
            "allow_clarification": os.getenv("ALLOW_CLARIFICATION", "true").lower() == "true",
            "max_concurrent_research_units": int(os.getenv("MAX_CONCURRENT_RESEARCH_UNITS", "5")),
            "search_api": os.getenv("SEARCH_API", "tavily"),
            "max_researcher_iterations": int(os.getenv("MAX_RESEARCHER_ITERATIONS", "6")),
            "max_react_tool_calls": int(os.getenv("MAX_REACT_TOOL_CALLS", "10")),
            "summarization_model": os.getenv("SUMMARIZATION_MODEL", "openai:gpt-4.1-mini"),
            "summarization_model_max_tokens": int(os.getenv("SUMMARIZATION_MODEL_MAX_TOKENS", "8192")),
            "max_content_length": int(os.getenv("MAX_CONTENT_LENGTH", "50000")),
            "research_model": os.getenv("RESEARCH_MODEL", "openai:gpt-4.1"),
            "research_model_max_tokens": int(os.getenv("RESEARCH_MODEL_MAX_TOKENS", "10000")),
            "compression_model": os.getenv("COMPRESSION_MODEL", "openai:gpt-4.1"),
            "compression_model_max_tokens": int(os.getenv("COMPRESSION_MODEL_MAX_TOKENS", "8192")),
            "final_report_model": os.getenv("FINAL_REPORT_MODEL", "openai:gpt-4.1"),
            "final_report_model_max_tokens": int(os.getenv("FINAL_REPORT_MODEL_MAX_TOKENS", "10000")),
        }
    
    def get_user_config(self, user_id: str, session_config: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Get configuration for a specific user, merging defaults with session-specific settings."""
        config = self.default_config.copy()
        
        # Apply session-specific configuration if provided
        if session_config:
            config.update(session_config)
        
        # TODO: In the future, load user-specific preferences from database
        # user_preferences = await self._load_user_preferences(user_id)
        # config.update(user_preferences)
        
        return config
    
    def create_runnable_config(self, thread_id: str, user_id: str, session_config: Optional[Dict[str, Any]] = None) -> RunnableConfig:
        """Create RunnableConfig for LangGraph execution."""
        configurable = self.get_user_config(user_id, session_config)
        configurable["thread_id"] = thread_id
        
        return {
            "configurable": configurable
        }
    
    def get_api_keys(self) -> Dict[str, str]:
        """Get API keys for different models from environment."""
        return {
            "openai": os.getenv("OPENAI_API_KEY"),
            "anthropic": os.getenv("ANTHROPIC_API_KEY"),
            "tavily": os.getenv("TAVILY_API_KEY"),
        }


# Global configuration manager instance
config_manager = ResearchConfigManager()

File: lib/deepResearcher/researchConfig.ts
/**
 * Research configuration management with user-specific settings.
 */

import { RunnableConfig } from "./configuration";

export interface ResearchConfig {
  maxStructuredOutputRetries: number;
  allowClarification: boolean;
  maxConcurrentResearchUnits: number;
  searchApi: string;
  maxResearcherIterations: number;
  maxReactToolCalls: number;
  summarizationModel: string;
  summarizationModelMaxTokens: number;
  maxContentLength: number;
  researchModel: string;
  researchModelMaxTokens: number;
  compressionModel: string;
  compressionModelMaxTokens: number;
  finalReportModel: string;
  finalReportModelMaxTokens: number;
  [key: string]: any;
}

export class ResearchConfigManager {
  defaultConfig: ResearchConfig;

  constructor() {
    this.defaultConfig = this.loadDefaultConfig();
  }

  /** Load default configuration from environment variables */
  private loadDefaultConfig(): ResearchConfig {
    const getInt = (key: string, def: string) => parseInt(process.env[key] || def, 10);
    const getBool = (key: string, def: string) =>
      (process.env[key] || def).toLowerCase() === "true";
    const getStr = (key: string, def: string) => process.env[key] || def;

    return {
      maxStructuredOutputRetries: getInt("MAX_STRUCTURED_OUTPUT_RETRIES", "3"),
      allowClarification: getBool("ALLOW_CLARIFICATION", "true"),
      maxConcurrentResearchUnits: getInt("MAX_CONCURRENT_RESEARCH_UNITS", "5"),
      searchApi: getStr("SEARCH_API", "tavily"),
      maxResearcherIterations: getInt("MAX_RESEARCHER_ITERATIONS", "6"),
      maxReactToolCalls: getInt("MAX_REACT_TOOL_CALLS", "10"),
      summarizationModel: getStr("SUMMARIZATION_MODEL", "openai:gpt-4.1-mini"),
      summarizationModelMaxTokens: getInt("SUMMARIZATION_MODEL_MAX_TOKENS", "8192"),
      maxContentLength: getInt("MAX_CONTENT_LENGTH", "50000"),
      researchModel: getStr("RESEARCH_MODEL", "openai:gpt-4.1"),
      researchModelMaxTokens: getInt("RESEARCH_MODEL_MAX_TOKENS", "10000"),
      compressionModel: getStr("COMPRESSION_MODEL", "openai:gpt-4.1"),
      compressionModelMaxTokens: getInt("COMPRESSION_MODEL_MAX_TOKENS", "8192"),
      finalReportModel: getStr("FINAL_REPORT_MODEL", "openai:gpt-4.1"),
      finalReportModelMaxTokens: getInt("FINAL_REPORT_MODEL_MAX_TOKENS", "10000"),
    };
  }

  /** Get configuration for a specific user, merging defaults with session-specific settings */
  getUserConfig(
    userId: string,
    sessionConfig?: Record<string, any>
  ): Record<string, any> {
    const config = { ...this.defaultConfig };

    if (sessionConfig) {
      Object.assign(config, sessionConfig);
    }

    // TODO: Future implementation for loading user-specific preferences
    // const userPreferences = await this.loadUserPreferences(userId);
    // Object.assign(config, userPreferences);

    return config;
  }

  /** Create RunnableConfig for LangGraph execution */
  createRunnableConfig(
    threadId: string,
    userId: string,
    sessionConfig?: Record<string, any>
  ): RunnableConfig {
    const configurable = this.getUserConfig(userId, sessionConfig);
    configurable.threadId = threadId;

    return { configurable };
  }

  /** Get API keys for different models from environment */
  getApiKeys(): Record<string, string | undefined> {
    return {
      openai: process.env.OPENAI_API_KEY,
      anthropic: process.env.ANTHROPIC_API_KEY,
      tavily: process.env.TAVILY_API_KEY,
    };
  }
}

// Global instance
export const configManager = new ResearchConfigManager();
